### PR TITLE
keyvault: optimising the lookup/adding a cache for Key Vaults

### LINKED ISSUE #6196: keyvault: caching of metadata required in larger terraform configurations
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.23
provider.azurerm v2.2.0

### Affected Resource(s)

* `azurerm_key_vault`
* `azurerm_key_vault_secret`
* `azurerm_virtual_machine_scale_set_extension`

### Terraform Configuration Files

```hcl

resource "azurerm_virtual_machine_scale_set_extension" "vmss_ext_msi" {
  virtual_machine_scale_set_id = azurerm_virtual_machine_scale_set.vmss.id
  name                         = "ManagedIdentityWindowsExtension"
  publisher                    = "Microsoft.ManagedIdentity"
  type                         = "ManagedIdentityExtensionForWindows"
  type_handler_version         = "1.0"
  settings                     = "{\"port\": 50342}"
  auto_upgrade_minor_version   = true
}
resource "azurerm_key_vault_secret" "kv_secret_01" {
  name         = "mySecret"
  value        = var.my_value
  key_vault_id = azurerm_key_vault.kv.id
  content_type = "text/plain"

  tags = {
    name        = "My Parameters"
  }

  depends_on = [
    azurerm_key_vault.kv,
    azurerm_key_vault_access_policy.kv_some_policy,
  ]
}
resource "azurerm_key_vault" "kv" {
  name                            = "my-kv"
  resource_group_name             = azurerm_resource_group.rg.name
  location                        = azurerm_resource_group.rg.location
  tenant_id                       = var.tenant_id
  enabled_for_deployment          = true
  enabled_for_disk_encryption     = true
  enabled_for_template_deployment = true
  purge_protection_enabled        = true
  soft_delete_enabled             = true

  sku_name = "standard"

  network_acls {
    default_action = "Deny"
    bypass         = "AzureServices"

    ip_rules = ["***","***"]

    virtual_network_subnet_ids = [
      azurerm_subnet.sub1.id,
      azurerm_subnet.sub2.id,
      azurerm_subnet.sub3_deployment_machine.id,
    ]
  }

  tags = {
    name        = "my good kv"
  }
}
```

### Expected Behavior

Terraform/AzureRM should always able to create a plan. 

### Actual Behavior

Every "terraform plan" action occurs different errors but the original error is the same which is "Context Deadline Exceeded". Sometimes we can create a plan with the all same resources but today I tried more than 15 times and no luck. Please see below for some errors which I selected uniquely. This kind of error occurs multiple times and for different resources in every plan action. 

```hcl
Error: Error checking if key vault "/subscriptions/***/resourceGroups/***/providers/Microsoft.KeyVault/vaults/***" for Secret "***" in Vault at url "https://***.vault.azure.net/" exists: Error making Read request on KeyVault "***" (Resource Group "***"): keyvault.VaultsClient#Get: Failure sending request: StatusCode=429 -- Original Error: context deadline exceeded

Error: Error retrieving the Resource ID the Key Vault at URL "https://***.vault.azure.net/": Error making Read request on KeyVault "***" (Resource Group "***"): keyvault.VaultsClient#Get: Failure sending request: StatusCode=429 -- Original Error: context deadline exceeded

Error: Error retrieving Extension "ManagedIdentityWindowsExtension" (Virtual Machine Scale Set "***" / Resource Group "***"): compute.VirtualMachineScaleSetExtensionsClient#Get: Failure sending request: StatusCode=429 -- Original Error: context deadline exceeded

Error: Error retrieving the Resource ID the Key Vault at URL "https://***.vault.azure.net/": Error GetKeyVaultId unable to list Key Vaults keyvault.VaultsClient#List: Failure sending request: StatusCode=429 -- Original Error: context deadline exceeded

Error: Error making Read request on AzureRM App Service "***": web.AppsClient#Get: Failure sending request: StatusCode=429 -- Original Error: context deadline exceeded
```

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

* Totally it's refreshing more than 1200 resources while planning. We have almost 80 key vaults and they contain lots of secrets. So most of the errors are related to Key Vaults. 
* These resources are already created and exist on Azure. So I can't use timeouts for creation and deletion. Also read timeouts didn't work. I've added it about 450 resources. I guess we need auto-retry..
* If we use resource targeting for those resources which we see on errors, we can create the plan successfully. So resources and our connections are healthy.
* We are only passing the "-refresh=true" parameter for plan action. 
* I've also tried with a parallelism parameter but no luck.
* Trying to plan from both a VM on Azure and my laptop.

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR supersedes #7794 and #6866 and fixes #6196 by adding a cache for Key Vaults rather than listing these for each resource, which was intended as a short-term hack.

In addition this PR moves all of the Key Vault helpers to the `keyvault/client` package, meaning the locks can be private fields (and potentially reused in future if necessary).

Notably this PR switches to looking the Key Vaults up using a filter on the Resources API to identify the Resource Group, meaning we're no longer pulling the entire list of Key Vaults in the subscription (originally necessary since it's not possible to retrieve a Key Vault without it's resource group directly)

Supersedes #7794
Supersedes #6866
Fixes #6196

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2021-01-27 at 13 29 27

File: azurerm/internal/services/keyvault/key_vault_secret_resource.go [hunk 1]
```diff
@@ -88,25 +87,28 @@ func resourceKeyVaultSecret() *schema.Resource {
...
+	keyVaultId, err := parse.VaultID(d.Get("key_vault_id").(string))
+	if err != nil {
+		return err
+	}
 
-	keyVaultBaseUrl, err := azure.GetKeyVaultBaseUrlFromID(ctx, vaultClient, keyVaultId)
+	keyVaultBaseUrl, err := keyVaultsClient.BaseUriForKeyVault(ctx, *keyVaultId)
 	if err != nil {
-		return fmt.Errorf("Error looking up Secret %q vault url from id %q: %+v", name, keyVaultId, err)
+		return fmt.Errorf("Error looking up Secret %q vault url from id %q: %+v", name, *keyVaultId, err)
```
Comment (jackofallops) [inline azurerm/internal/services/keyvault/key_vault_secret_resource.go:105]:
```suggestion
		return fmt.Errorf("looking up Secret %q vault url from id %q: %+v", name, *keyVaultId, err)
```

File: azurerm/internal/services/keyvault/key_vault_secret_resource.go [hunk 2]
```diff
@@ -88,25 +87,28 @@ func resourceKeyVaultSecret() *schema.Resource {
...
 	}
 
-	existing, err := client.GetSecret(ctx, keyVaultBaseUrl, name, "")
+	existing, err := client.GetSecret(ctx, *keyVaultBaseUrl, name, "")
 	if err != nil {
 		if !utils.ResponseWasNotFound(existing.Response) {
-			return fmt.Errorf("Error checking for presence of existing Secret %q (Key Vault %q): %s", name, keyVaultBaseUrl, err)
+			return fmt.Errorf("Error checking for presence of existing Secret %q (Key Vault %q): %s", name, *keyVaultBaseUrl, err)
```
Comment (jackofallops) [inline azurerm/internal/services/keyvault/key_vault_secret_resource.go:111]:
```suggestion
			return fmt.Errorf("checking for presence of existing Secret %q (Key Vault %q): %s", name, *keyVaultBaseUrl, err)
```

File: azurerm/internal/services/keyvault/key_vault_secret_resource.go [hunk 3]
```diff
@@ -257,9 +264,9 @@ func resourceKeyVaultSecretUpdate(d *schema.ResourceData, meta interface{}) erro
 	}
 
 	// "" indicates the latest version
-	read, err2 := client.GetSecret(ctx, id.KeyVaultBaseUrl, id.Name, "")
-	if err2 != nil {
-		return fmt.Errorf("Error getting Key Vault Secret %q : %+v", id.Name, err2)
+	read, err := client.GetSecret(ctx, id.KeyVaultBaseUrl, id.Name, "")
+	if err != nil {
+		return fmt.Errorf("Error getting Key Vault Secret %q : %+v", id.Name, err)
```
Comment (jackofallops) [inline azurerm/internal/services/keyvault/key_vault_secret_resource.go:269]:
```suggestion
		return fmt.Errorf("getting Key Vault Secret %q : %+v", id.Name, err)
```

**Review by jackofallops — APPROVED** [review]
Thanks @tombuildsstuff - Couple of suggested changes for error messages while we're touching those lines anyway, but otherwise LGTM 👍

**Pull request merged**